### PR TITLE
[CUDA] Add non-overlapping fast path for avg_pool2d backward

### PR DESCRIPTION
**Pull request by Ansh-Karnwal**
### Summary

Adds a dedicated CUDA backward kernel for `avg_pool2d` when the pooling windows tile the input exactly (`stride == kernel_size`, no padding, no `ceil_mode`) -- the common downsampling case such as `avg_pool2d(2, 2)`. In that regime each input element maps to exactly one output window with a constant divisor, so the general kernel's per-element window scan and per-window bound/divisor recomputation collapse to a single load and a single divide.

All other configurations are routed to the existing general kernel unchanged. The fast kernels omit `count_include_pad` because it cannot affect the result when there is no padding.

### Why

The general backward kernel is instruction-count bound in this case (measured ~61% SM / ~22% DRAM on an RTX 3080), spending most of its integer ALU work recomputing window bounds and divisors that are constant for non-overlapping windows.

### Numerics

The fast path is numerically identical to the general kernel -- bitwise-equal across float/double/half/bfloat16 in testing. This is purely a performance change, not a semantics change.

### Performance (RTX 3080, sm_86, backward only)

Same-binary A/B (`ceil_mode=True` forces the general kernel on a divisible input; identical work, only the kernel differs):

| shape | k | dtype | general | fast | speedup |
|-------|---|-------|---------|------|---------|
| 64x64x224x224 | 2 | float32 | 7.06 ms | 5.10 ms | 1.39x |
| 64x64x224x224 | 2 | float16 | 7.26 ms | 4.55 ms | 1.59x |
| 64x64x224x224 | 2 | bfloat16 | 6.97 ms | 4.57 ms | 1.53x |
| 32x64x224x224 | 4 | float32 | 3.43 ms | 2.44 ms | 1.41x |
| 128x64x56x56 | 2 | float32 (NHWC) | 0.95 ms | 0.69 ms | 1.38x |

Geomean 1.43x (min 1.38x, max 1.59x) across the configs tested.

### Binary size

The change adds 16 new device-kernel instantiations (dtype x int/long index x NCHW/NHWC), mirroring the existing 16 general-backward instantiations. On sm_86 they total ~186 KB of SASS, roughly 46% of the general-backward kernels' ~402 KB (each fast kernel is smaller since it drops the window loop). This scales with the number of built GPU archs, like any kernel.

### Tests

- `test_avg_pool2d_backward_nonoverlapping`: device-generic, compares the fast path against an independent reference over both memory formats, non-divisible shapes (dropped rows/cols), asymmetric kernels, single batch/channel, and `divisor_override`.
- `test_avg_pool2d_backward_nonoverlapping_large`: `@largeTensorTest("18GB")`, exercises the 64-bit index path (`numel > INT_MAX`).
- Existing `avg_pool2d` pooling + OpInfo gradcheck tests pass unchanged.
- `compute-sanitizer memcheck`: 0 errors across all fast-path branches.

---
<sub>PR description formatted with AI for readability.</sub>

### CODE REVIEW

**Comment by Skylion007** [comment]
@claude is there no way to optimize this fast path into the existing kernels or reduce or otherwise simplify it to get similar benefits without template bloat?

**Comment by Ansh-Karnwal** [comment]
Thanks, I measured this. Folding the fast body into the existing kernel behind a runtime flag only recovers a small part of the win (RTX 3080, all bitwise-identical).

| 64x64x224x224, k=2 | general (main) | in-place hoist | dedicated |
|---|---|---|---|
| speedup | 1.00x | 1.13x | 1.61x |

I found that the dominant cost was not the divisor recompute or clamps and found that it was the `phstart/phend` bound math and the loop, which only `ph = h/kH` collapses.

On the bloat, I managed to cut it down.

1. Merged the NCHW and NHWC fast kernels into one with a grid-uniform `channels_last` flag (no divergence). That's 16 instantiations down to 8, bitwise-identical, no regression (still 1.44x geomean). The SASS doesn't fully halve since each unified kernel carries both decode paths (~142 KB per arch for the 8, roughly a 20 to 25% cut). Also puts the `channels` arg you flagged to use in the channels_last path.
2. Took the divisor hoist for the general kernel anyway, behind a host `uniform_divisor` flag, no new instantiations. 1.1 to 1.2x on the overlapping and padded configs the fast path skips (`avg_pool2d(3, stride=2)`, padded pooling).

**Comment by Ansh-Karnwal** [comment]
@Skylion007

File: aten/src/ATen/native/cuda/AveragePool2d.cu [hunk 1]
```diff
@@ -238,6 +237,48 @@ __global__ void avg_pool2d_backward_out_cuda_frame_nhwc(const index_t nthreads,
...
+  const int divide_factor = use_divisor ? divisor_override : kernel_h * kernel_w;
+  CUDA_KERNEL_LOOP_TYPE(index, nthreads, index_t) {
+    int w, h;
+    index_t slice_base;
+    int64_t inner_stride;
+    if (channels_last) {
+      const int c = index % channels;
+      w = (index / channels) % width;
+      h = (index / channels / width) % height;
+      const index_t n = index / channels / width / height;
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/AveragePool2d.cu]:
```suggestion
      h = (index / (channels * width)) % height;
      const index_t n = index / (channels * width * height);

File: aten/src/ATen/native/cuda/AveragePool2d.cu [hunk 2]
```diff
@@ -238,6 +237,48 @@ __global__ void avg_pool2d_backward_out_cuda_frame_nhwc(const index_t nthreads,
...
+      slice_base = n * channels * pooled_height * pooled_width + c;
+      inner_stride = channels;
+    } else {
+      w = index % width;
+      h = (index / width) % height;
+      const index_t nc = index / width / height;  // n * channels + c
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/AveragePool2d.cu]:
```suggestion
      const index_t nc = index / (width * height);  // n * channels + c
```
Integer division is expensive...

File: aten/src/ATen/native/cuda/AveragePool2d.cu [hunk 3]
```diff
@@ -238,6 +237,48 @@ __global__ void avg_pool2d_backward_out_cuda_frame_nhwc(const index_t nthreads,
...
+    const int divisor_override, bool use_divisor, bool channels_last) {
+  const int divide_factor = use_divisor ? divisor_override : kernel_h * kernel_w;
+  CUDA_KERNEL_LOOP_TYPE(index, nthreads, index_t) {
+    int w, h;
+    index_t slice_base;
+    int64_t inner_stride;
+    if (channels_last) {
+      const int c = index % channels;
+      w = (index / channels) % width;
+      h = (index / channels / width) % height;
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/AveragePool2d.cu]:
```suggestion
      h = (index / (channels * width)) % height;
```

Feels liek channels * width could be cached too...

**Comment by Ansh-Karnwal** [comment]
@Skylion007 The literal version measured slower. `index / (channels * width)` is a fresh division that doesn't reuse the `index / channels` quotient the `w` line already computes, so it adds a division instead of removing one (0.77 to 1.0x in a back to back microbench).

Reusing the intermediate quotients is as fast or faster with no added divisions, compute `index / channels` once, take `w`, divide by width for `h`, then by height for `n`. End to end it lands within run to run noise since the decode is a small slice of the kernel, but it doesn't regress. Pushed.

**Comment by Skylion007** [comment]
Can these optimizations be applied to the 3D kernel too minus the non overlapping path?

**Review by Skylion007 — APPROVED** [review]

**Comment by Ansh-Karnwal** [comment]
@Skylion007 The divisor hoist ports to the general 3D backward behind the same host `uniform_divisor` flag. The clamp skip doesn't, since 3D backward is a per-output scatter, not the 2D per-input gather, so the clamped bounds are the scatter loop bounds and stay live. Hoisting the uniform divisor still drops the per-thread `pool_size` compute and branch. Measured on a 3080, bitwise-identical:

| path | speedup |
|---|---|
| overlapping k=3 s=1 p=1 (atomicAdd) | 1.0 to 1.08x, shape dependent |
| non-overlap padded k=2 s=2 p=1 (direct write) | ~1.02x |

Smaller than the 2D win since 3D backward is scatter and atomic bound, not instruction bound. Can add it in a follow-up if you want the generality, same host-flag change with no new instantiations.

**Comment by Skylion007** [comment]
We should add a follow up PR for that.

**Review by Skylion007 — APPROVED** [review]

**Comment by Ansh-Karnwal** [comment]
On it.

**Comment by Ansh-Karnwal** [comment]
Pushed a test only fix in 521abfd9a37, the kernel is unchanged. The merge run surfaced three things and none were in the kernel itself, every CUDA correctness test passed in that shard.

- The two clang20 builds failed on a runner pod that never came online, an infra flake.
- The large 64 bit index test OOMed. `(grad_input != 0).sum()` upcasts a 2.1B element bool mask to int64 for the reduction, a 17GB allocation on top of the half tensors. Replaced it with trailing slice `count_nonzero` checks that only touch the dropped region.
- The device generic test also runs on MPS, where avg_pool2d backward diverges from the CPU and CUDA reference (both of which match). Marked it `@expectedFailureMPS` so it stays a tripwire if MPS is ever fixed instead of silently skipping.

**Comment by Skylion007** [comment]
@claude how would you properly fix the MPS issues?

**Comment by Ansh-Karnwal** [comment]
@Skylion007 Added `@dtypesIfMPS(torch.float)` so double is out of MPS scope, matching test_avg_pool3d_nhwc. The remaining float32 failure is the channels_last path, the same NHWC backend bug that test_avg_pool2d_nhwc already xfails, so I pointed the comment there instead of leaving it vague. Pushed in 4e55dc957a5.

The 3D uniform_divisor hoist is on my list for a follow-up. The Metal avg_pool_backward 2D specialization is the proper fix to get 2D backward off MPSGraph and clear both xfails, but it is a bigger lift since I would build and validate it on my Mac. Happy to take it if you want it in scope, otherwise it may be a better fit for someone already set up on MPS. The xfails point at the tracked NHWC bug either way.

File: test/nn/test_pooling.py [hunk 4]
```diff
@@ -1173,6 +1173,104 @@ def helper(
...
+            xc = x.clone().requires_grad_()
+            xg = x.to(device=device)
+            if channels_last:
+                xg = xg.contiguous(memory_format=torch.channels_last)
+            xg = xg.requires_grad_()
+            F.avg_pool2d(xc, **kw).sum().backward()
+            F.avg_pool2d(xg, **kw).sum().backward()
+            self.assertEqual(xg.grad, xc.grad)
+
+        for channels_last in (False, True):
```
Comment (Skylion007) [inline test/nn/test_pooling.py]: Nit: can this be lifted into a parameterize block so we get better failure marks? Also would clean this up quite a bit

**Comment by Skylion007** [comment]
@claude show how the unit test here could be improved with parameterize

**Comment by Ansh-Karnwal** [comment]
@Skylion007 Done, parametrized both tests in 4992e4f830f. Each config is its own subtest now, so failures name the exact case like `test_avg_pool2d_backward_nonoverlapping_channels_last_dropped_row_col_cuda_float16`, and one bad config no longer masks the rest. The channels_last axis is a real parametrization, so the xfail scopes to just the channels_last=True subtests and the contiguous MPS path becomes an enforced pass instead of getting swallowed.

**Comment by Skylion007** [comment]
@claude review this PR

File: test/nn/test_pooling.py [hunk 5]
```diff
@@ -1169,6 +1174,135 @@ def helper(
...
+        self.assertEqual(covered.min().item(), expected)
+        self.assertEqual(covered.max().item(), expected)
+        # Everything outside the covered region (the trailing dropped row, since
+        # input height 32785 is not a multiple of 16) must be zero. Check the
+        # trailing slices directly rather than counting nonzeros over the whole
+        # grad_input, which would materialize a full-size int64 mask (OOM).
+        self.assertEqual(grad_input[..., ph * k :, :].count_nonzero().item(), 0)
+        self.assertEqual(grad_input[..., :, pw * k :].count_nonzero().item(), 0)
+
+    @onlyCUDA
```
Comment (Skylion007) [inline test/nn/test_pooling.py:1459]: Is this onlyCUDA versus skip MPS I guess lol?

**Review by Skylion007 — APPROVED** [review]

**Comment by pbielak** [comment]
@Ansh-Karnwal please rebase your PR - I just merged https://github.com/pytorch/pytorch/pull/181571 which updated the `test_pooling.py` file

**Comment by Ansh-Karnwal** [comment]
I was out for a little, catching up

@pbielak Rebased onto main in fdcd599, picks up #181571.

@Skylion007 Also fixed the MPS XPASS that blocked the last merge. The channels_last xfail was firing on the single_batch_channel config, which has C=1 where channels_last is a no-op, so MPS computes it correctly and the strict xfail turns that pass into a failure. Since the MPS failure is not uniform across the parametrized shapes, I switched channels_last from expectedFailureMPS to skipMPS with a comment on why. The known NHWC backend bug is still tracked by test_avg_pool2d_nhwc.

**Comment by Ansh-Karnwal** [comment]
https://github.com/pytorch/pytorch/pull/195119

got around to the 3D as well

**Comment by Ansh-Karnwal** [comment]
@Skylion007 mind kicking off CI here? the rebase re-parked the workflows at action_required and i can't approve my own. nothing changed since your approval apart from the MPS XPASS that blocked the last merge, fixed in fdcd599.

also #195119 is the 3d hoist you asked for, whenever you have a minute.

File: test/nn/test_pooling.py [hunk 6]
```diff
@@ -1374,6 +1376,137 @@ def helper(
...
+    @dtypesIfCUDA(torch.half, torch.bfloat16, torch.float, torch.double)
+    @dtypesIfMPS(torch.float)
+    @parametrize_test(
+        "channels_last",
+        [
+            subtest(False, name="contiguous"),
+            # MPS channels_last avg_pool2d backward is broken (see test_avg_pool2d_nhwc)
+            # for C > 1 but computes the degenerate C == 1 configs correctly, so the
+            # failure is not uniform across the parametrized shapes; skip rather than xfail.
+            subtest(True, name="channels_last", decorators=[skipMPS]),
```
Comment (Skylion007) [inline test/nn/test_pooling.py:1389]: Should this be 
```
            subtest(True, name="channels_last", decorators=[xfailIfMPS]),

Comment (Ansh-Karnwal) [reply test/nn/test_pooling.py:1389]: `xfailIfMPS` is strict `expectedFailure` on MPS, same as `expectedFailureMPS` which I had here. It XPASS-failed on `single_batch_channel`. C=1 makes channels_last a no-op, so MPS gets it right and strict xfail flags the pass. The failure isn't uniform across the shapes, so it has to be skip.
